### PR TITLE
[styled-components] Fix union type of props being lost

### PR DESCRIPTION
**Pull request by eps1lon**
Components wrapped in `styled` currently loose any union type information about the props of the original component.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes:
  - This is an internal change to the typings
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by VincentLanglet** [comment]
@eps1lon There is a lot of package using the same Omit definition.
`type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;`

Does this means we'll have the same issue for others libraries ?
https://github.com/DefinitelyTyped/DefinitelyTyped/search?q=Omit&unscoped_q=Omit

**Comment by eps1lon** [comment]
@VincentLanglet I believe so. See Microsoft/TypeScript#28339 for more context.

File: types/styled-components/index.d.ts [hunk 1]
```diff
@@ -345,8 +346,10 @@ export type ThemedCssFunction<T extends object> = BaseThemedCssFunction<
 >;
 
 // Helper type operators
-type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
-type WithOptionalTheme<P extends { theme?: T }, T> = Omit<P, "theme"> & {
+// Pick that distributes over union types
+export type PickU<T, K extends keyof T> = T extends any ? {[P in K]: T[P]} : never;
+export type OmitU<T, K extends keyof T> = T extends any ? PickU<T, Exclude<keyof T, K>> : never;
```
Comment (Jessidhia) [inline types/styled-components/index.d.ts]: Interesting that `T extends any` works around the issue 🤔

**Review by Jessidhia — APPROVED** [review]

File: types/styled-components/index.d.ts (hunk 1 above)
Comment (VincentLanglet) [reply types/styled-components/index.d.ts]: @Jessidhia The solution came from your issue [here](https://github.com/Microsoft/TypeScript/issues/28339)

And it's explained : 
> For example, an instantiation of `T extends U ? X : Y` with the type argument `A | B | C` for `T` is resolved as `(A extends U ? X : Y) | (B extends U ? X : Y) | (C extends U ? X : Y)`.

Here : https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#distributive-conditional-types

**Pull request merged**